File: typ2anki/cardscache.py
from dataclasses import dataclass
from typing import Dict, Tuple
from .api import (
    get_cards_cache_string,
    upload_media,
    CARDS_CACHE_FILENAME,
)
import json
from .card_wrapper import CardInfo
from .config import config
from .utils import hash_string

CACHE_HASH_PART_LENGTH = 34


def cache_concat_hashes_padding(hash1: str, hash2: str) -> str:
    return hash1[:CACHE_HASH_PART_LENGTH].ljust(
        CACHE_HASH_PART_LENGTH, "0"
    ) + hash2[:CACHE_HASH_PART_LENGTH].ljust(CACHE_HASH_PART_LENGTH, "0")


class CardsCacheManager:
    current_ankiconf_hash = None
    current_config_hash = None
    static_hash = None
    current_card_hashes: Dict[str, str] = {}
    cache: Dict[str, str] = {}
    ignore_config_change = False

    def __init__(self):
        self.load_cache()

    # Creating a static_hash to be used as a base for the file hashes: objective of this is to be able to use the same cache file for different configurations
    def add_static_hashes(self, ankiconf_hash, config_hash):
        self.current_ankiconf_hash = ankiconf_hash
        self.current_config_hash = config_hash
        self.static_hash = hash_string(f"{ankiconf_hash}{config_hash}")

    def load_cache(self):
        # If we are not checking checksums, clear the cache
        if not config().check_checksums:
            print("Not checking checksums, using an empty cache")
            self.cache = {}
            return
        s = get_cards_cache_string()
        if not s:
            self.cache = {}
        else:
            try:
                self.cache = json.loads(s)
            except Exception as e:
                print("Error loading cache")

        if self.cache is None:
            self.cache = {}

    def add_current_card_hash(self, deck_name, card_id, card_hash):
        assert (
            self.static_hash is not None
        ), "Static hash is not set. Call add_static_hashes first."
        self.current_card_hashes[f"{deck_name}_{card_id}"] = (
            cache_concat_hashes_padding(self.static_hash, card_hash)
        )

    # Returns: (changed_configs,total_cards)
    def has_config_changed(self) -> Tuple[int, int]:
        config_changes = 0
        total_cards = 0
        for key, cached_hash in self.cache.items():
            if key in self.current_card_hashes:
                total_cards += 1
                if (
                    cached_hash[:CACHE_HASH_PART_LENGTH]
                    != self.current_card_hashes[key][:CACHE_HASH_PART_LENGTH]
                ):
                    config_changes += 1
        return (config_changes, total_cards)

    def detect_config_change(self):
        if not config().check_checksums:
            return

        if config().recompile_on_config_change != "_":
            self.ignore_config_change = (
                config().recompile_on_config_change != "y"
            )
            return

        config_changes, total_cards = self.has_config_changed()
        if config().dry_run:
            print(
                f"Config changes detected: {config_changes}, Total cards already cached: {total_cards}"
            )

        if total_cards == 0:
            return

        if config_changes / total_cards > 0.5:
            try:
                inp = input(
                    "A configuration or ankiconf.typ change has been detected. Do you wish to recompile all cards with this new config? (Y/n): "
                ).lower()
            except EOFError:
                print("No input available. Assuming 'Y'.")
                inp = "y"

            if inp == "n" or inp == "no":
                self.ignore_config_change = True
            else:
                self.ignore_config_change = False

    def remove_card_hash(self, card: CardInfo):
        id = f"{card.deck_name}_{card.card_id}"
        self.current_card_hashes.pop(id, None)
        self.cache.pop(id, None)

    def card_needs_push(self, card: CardInfo):
        if config().check_checksums == False:
            return True
        id = f"{card.deck_name}_{card.card_id}"

        new_hash = self.current_card_hashes.get(id, "notfound")
        old_hash = self.cache.get(id, "notfound")

        if new_hash == "notfound" or old_hash == "notfound":
            return True

        if self.ignore_config_change:
            return (
                new_hash[-CACHE_HASH_PART_LENGTH:]
                != old_hash[-CACHE_HASH_PART_LENGTH:]
            )
        return new_hash != old_hash

    def save_cache(self, output_path):
        temp_file = output_path / CARDS_CACHE_FILENAME
        new_cache = self.cache
        new_cache.update(self.current_card_hashes)

        with open(temp_file, "w") as file:
            file.write(json.dumps(new_cache))

        upload_media(temp_file)
        temp_file.unlink(missing_ok=True)

File: typ2anki-rust/src/cards_cache.rs
use std::collections::HashMap;

use crate::output::{OutputManager, OutputMessage};
use crate::utils::{self, hash_string};
use crate::{anki_api, config};

const CACHE_HASH_PART_LENGTH: usize = 34;

#[derive(Debug, Clone)]
pub struct CardsCacheManager {
    pub static_hash: String,
    pub old_cache: HashMap<String, String>,
    pub new_cache: HashMap<String, String>,
}

pub fn card_key(deck_name: &str, card_id: &str) -> String {
    format!("{}_{}", deck_name, card_id)
}

fn cache_concat_hashes_padding(hash1: &str, hash2: &str) -> String {
    let mut out = String::new();
    out.push_str(hash1);
    out.push_str(&"0".repeat(CACHE_HASH_PART_LENGTH.saturating_sub(hash1.len())));
    out.push_str(&"0".repeat(CACHE_HASH_PART_LENGTH.saturating_sub(hash2.len())));
    out.push_str(hash2);
    out
}

impl CardsCacheManager {
    pub fn init(ankiconf_hash: String, _output: &OutputManager) -> Self {
        let cfg = config::get();
        let static_hash =
            hash_string(format!("{}{}", ankiconf_hash, cfg.config_hash.as_ref().unwrap()).as_str());
        let cache: HashMap<String, String>;

        if !cfg.check_checksums {
            cache = HashMap::new();
        } else {
            let s = anki_api::get_cards_cache_string().unwrap_or("{}".to_string());
            cache = serde_json::from_str(&s).unwrap_or(HashMap::new());
        }

        Self {
            static_hash,
            new_cache: HashMap::new(),
            old_cache: cache,
        }
    }

    pub fn add_card_hash(&mut self, deck_name: &str, card_id: &str, content_hash: &str) {
        self.new_cache.insert(
            card_key(deck_name, card_id),
            cache_concat_hashes_padding(&self.static_hash, content_hash),
        );
    }

    // Removes the new hash for a card (used when a card fails to compile/upload)
    pub fn remove_card_hash(&mut self, deck_name: &str, card_id: &str) {
        let key = card_key(deck_name, card_id);
        self.new_cache.remove(&key);
        self.old_cache.remove(&key);
    }

    pub fn detect_configuration_change(&mut self, output: &OutputManager) {
        let cfg = config::get();
        if !cfg.check_checksums {
            return;
        }

        let mut config_changes = 0;
        let mut total_cards = 0;
        for (k, v) in &self.old_cache {
            total_cards += 1;
            if let Some(new_v) = self.new_cache.get(k) {
                if v[..CACHE_HASH_PART_LENGTH] != new_v[..CACHE_HASH_PART_LENGTH] {
                    config_changes += 1;
                }
            }
        }

        if cfg.dry_run {
            output.send(OutputMessage::DbgConfigChangeDetection {
                total_cards,
                config_changes,
            });
        }

        if cfg.recompile_on_config_change.read().unwrap().is_none() {
            if total_cards > 0 && (config_changes as f64) / (total_cards as f64) >= 0.2 {
                if output.ask_yes_no("A configuration or ankiconf.typ change has been detected. Do you wish to recompile all cards with this new config? (Y/n)") {
                    *cfg.recompile_on_config_change.write().unwrap() = Some(true);
                } else {
                    *cfg.recompile_on_config_change.write().unwrap() = Some(false);
                }
            }
        }
    }

    pub fn save_cache(&self, output: &OutputManager) {
        let push: HashMap<String, String> = self
            .old_cache
            .clone()
            .into_iter()
            .chain(self.new_cache.clone().into_iter())
            .collect();
        let s = serde_json::to_string(&push).unwrap_or("{}".to_string());
        let payload = utils::b64_encode(s);
        if let Err(e) = anki_api::upload_file(anki_api::CARDS_CACHE_FILENAME.into(), &payload) {
            output.send(OutputMessage::ErrorSavingCache(e));
        } else {
            output.send(OutputMessage::DbgSavedCache);
        }
    }
}
